backbone/src/main/java/org/researchstack/backbone/step/active/recorder/AudioRecorder.java

```diff
@@ -0,0 +1,288 @@
+
+import org.researchstack.backbone.R;
+import org.researchstack.backbone.result.AudioResult;
+import org.researchstack.backbone.step.Step;
+import org.researchstack.backbone.utils.LogExt;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Date;
+
+/**
+ * Created by TheMDP on 2/26/17.
+ *
+ * The AudioRecorder records audio in a format specified by the AudioRecorderSettings
+ * It bundles the result in an AudioResult object in which it returns in the recorder listener
+ */
+
+public class AudioRecorder extends Recorder {
+
+    private static final double VOLUME_CLAMP_IN_DECIBELS = 60.0;
+    public static final int MAX_VOLUME = 32767;
+
+    /**
+     * This is the duration in between checking for the max sample amplitude of the audio recorder
+     * As of now, 100 ms is sufficient to check if the audio is "too loud"
+     */
+    private static final int AVERAGE_MAX_VOLUME_DURATION = 100;
+
+    private AudioRecorderListener audioRecorderListener;
+
+    /**
+     * Used to check amplitude of the sample at a desired frequency
+     */
+    private Handler  mainHandler;
+    private Runnable sampleMonitorRunnable;
+
+    /**
+     * The AudioRecorder can give average sample callbacks at any desired frequency
+     * These variables keep track of the running average per callback window
+     */
+    private long sampleSumSinceLastCallback;
+    private int samplesSinceLastCallback;
+    private long timeOfLastCallback;
+    private long msBetweenCallbacks;
+
+    /**
+     * These keep track of the entire rolling average for the entire life-cycle of AudioRecorder
+     * It can be used to get an overall picture of the audio's background noise
+     */
+    private long totalRollingAvg;
+    private int totalRollingAvgSampleCount;
+
+    private AudioRecorderSettings settings;
+    private MediaRecorder mediaRecorder;
+    private long startTime;
+
+    AudioRecorder(AudioRecorderSettings settings,
+                  String identifier,
+                  Step step,
+                  File outputDirectory)
+    {
+        super(identifier, step, outputDirectory);
+        this.settings = settings;
+        mainHandler = new Handler();
+    }
+
+    @Override
+    public void start(Context context) {
+        if (mediaRecorder != null) {
+            throw new IllegalStateException("Cannot start media recorder since it has already been started");
+        }
+
+        if (settings == null) {
+            throw new IllegalStateException("Cannot start media recorder since settings is null");
+        }
+
+        // In-app permissions were added in Android 6.0
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            PackageManager pm = context.getPackageManager();
+            int hasPerm = pm.checkPermission(Manifest.permission.RECORD_AUDIO, context.getPackageName());
+            if (hasPerm != PackageManager.PERMISSION_GRANTED) {
+                if (getRecorderListener() != null) {
```

Comment:
You fail if permission has not been granted as opposed to asking for permission?